### PR TITLE
rbd: enable expand operation for intree volumes

### PR DESCRIPTION
**Pull request by humblec**
This commit enable the resize operation [1] for in-tree volumes.
new helper has been introduced here to aid the enablement or to
make it clean with existing code base.

[1] https://github.com/ceph/ceph-csi/blob/devel/docs/design/proposals/intree-migrate.md?plain=1#L66

Updates #2509

Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

Additional Info:
- [x] no unit test added for the helper newVolFromVolIDWithMigration() as it contains cr format as in other left out helpers.
- [x] no e2e as it need support from sidecar..etc.
- [x] I have tested this manually and the resize has worked as expected and also for other operations.

Here is the log: 
```
I1222 06:34:26.791024 1000552 utils.go:177] ID: 26 Req-ID: mig_mons-b7f67366bb43f32e07d8a261a7840da9_image-773226a6-dbc3-47b5-b47c-bfa8d64b1e2e_7265706c696361706f6f6c GRPC call: /csi.v1./ExpandVolume
I1222 06:34:26.791314 1000552 utils.go:181] ID: 26 Req-ID: mig_mons-b7f67366bb43f32e07d8a261a7840da9_image-773226a6-dbc3-47b5-b47c-bfa8d64b1e2e_7265706c696361706f6f6c GRPC request: {"capacity_range":{"required_bytes":4294967296},....
I1222 06:34:26.828325 1000552 mount_linux.go:487] Attempting to determine if disk "/dev/rbd0" is formatted using blkid with args: ([-p -s TYPE -s PTTYPE -o export /dev/rbd0])
I1222 06:34:26.830157 1000552 mount_linux.go:490] Output: "DEVNAME=/dev/rbd0\nTYPE=ext4\n"
I1222 06:34:26.830187 1000552 resizefs_linux.go:56] ResizeFS.Resize - Expanding mounted volume /dev/rbd0
I1222 06:34:26.832030 1000552 resizefs_linux.go:71] Device /dev/rbd0 resized successfully
I1222 06:34:26.832140 1000552 utils.go:188] ID: 26 Req-ID: mig_mons-b7f67366bb43f32e07d8a261a7840da9_image-773226a6-dbc3-47b5-b47c-bfa8d64b1e2e_7265706c696361706f6f6c GRPC response: {}
...


[hchiramm@hchiramm ]$ mount |grep rbd
/dev/rbd0 on /var/lib/kubelet/plugins/kubernetes.io/csi/pv/pvc-c827c204-94e7-4ed9-87e9-fce212fe61a0/globalmount/mig_mons-b7f67366bb43f32e07d8a261a7840da9_image-773226a6-dbc3-47b5-b47c-bfa8d64b1e2e_7265706c696361706f6f6c type ext4 (rw,relatime,stripe=16,_netdev)
/dev/rbd0 on /var/lib/kubelet/pods/80dc68c4-049d-49cb-8933-abf64c35bb65/volumes/kubernetes.io~csi/pvc-c827c204-94e7-4ed9-87e9-fce212fe61a0/mount type ext4 (rw,relatime,stripe=16,_netdev)
[hchiramm@hchiramm ]$ kubectl exec -ti in-task-pv-pod -- df -kh|grep nginx
/dev/rbd0                                4.0G  4.0M  3.9G   1% /usr/share/nginx/html
[hchiramm@hchiramm ]$ 
```

### CODE REVIEW

File: internal/rbd/migration.go [hunk 1]
```diff
@@ -77,35 +77,43 @@ func parseMigrationVolID(vh string) (*migrationVolID, error) {
 // deleteMigratedVolume get rbd volume details from the migration volID
 // and delete the volume from the cluster, return err if there was an error on the process.
 func deleteMigratedVolume(ctx context.Context, parsedMigHandle *migrationVolID, cr *util.Credentials) error {
+	rv, err := genVolFromMigVolID(ctx, parsedMigHandle, cr)
+	if err != nil {
+		return err
+	}
+	err = deleteImage(ctx, rv, cr)
+	if err != nil {
+		log.ErrorLog(ctx, "failed to delete rbd image : %s, err: %v", rv.RbdImageName, err)
```
Comment (Madhu-1) [inline internal/rbd/migration.go]: log rv not just the image name

File: internal/rbd/controllerserver.go [hunk 2]
```diff
@@ -1447,9 +1447,7 @@ func (cs *ControllerServer) ControllerExpandVolume(
 		return nil, status.Error(codes.InvalidArgument, err.Error())
 	}
 	defer cr.DeleteCredentials()
-
-	rbdVol, err := GenVolFromVolID(ctx, volID, cr, req.GetSecrets())
-	defer rbdVol.Destroy()
```
Comment (Madhu-1) [inline internal/rbd/controllerserver.go:1452]: what is the reason for this change?

File: internal/rbd/migration.go [hunk 3]
```diff
@@ -77,35 +77,43 @@ func parseMigrationVolID(vh string) (*migrationVolID, error) {
 // deleteMigratedVolume get rbd volume details from the migration volID
 // and delete the volume from the cluster, return err if there was an error on the process.
 func deleteMigratedVolume(ctx context.Context, parsedMigHandle *migrationVolID, cr *util.Credentials) error {
+	rv, err := genVolFromMigVolID(ctx, parsedMigHandle, cr)
+	if err != nil {
+		return err
+	}
```
Comment (Madhu-1) [inline internal/rbd/migration.go:83]: @humblec have you tested these changes manually?

File: internal/rbd/rbd_util.go [hunk 4]
```diff
@@ -2071,3 +2071,27 @@ func strategicActionOnLogFile(ctx context.Context, logStrategy, logFile string)
 		log.ErrorLog(ctx, "unknown cephLogStrategy option %q: hint: 'remove'|'compress'|'preserve'", logStrategy)
 	}
 }
+
+// genVolFromVolIDWithMigration populate a rbdVol structure based on the volID format.
+func genVolFromVolIDWithMigration(
```
Comment (Madhu-1) [inline internal/rbd/rbd_util.go:2076]: function name says this is for migration but you are generating rbdVol for both cases.

**Review by Madhu-1 — CHANGES_REQUESTED** [review]
if not available Please add e2e changes for this one.

File: internal/rbd/controllerserver.go (hunk 2 above)
Comment (humblec) [reply internal/rbd/controllerserver.go:1452]: madhu, just the parsing and volID creation is in the helper.

File: internal/rbd/migration.go (hunk 3 above)
Comment (humblec) [reply internal/rbd/migration.go:83]: @Madhu-1 Thats what in the PR description?

File: internal/rbd/rbd_util.go (hunk 4 above)
Comment (humblec) [reply internal/rbd/rbd_util.go:2076]: it is same model as what @nixpanic suggested for credential support , similar one to @newUserCredentialsWithMigration to make it clean.Otherwise a if ...else in the caller does the job.

File: internal/rbd/migration.go (hunk 3 above)
Comment (Madhu-1) [reply internal/rbd/migration.go:83]: Have you tested the delete workflow the rv connect and destroy is done in genVolFromMigVolID but the connection is needed for deleteImage to work

File: internal/rbd/controllerserver.go (hunk 2 above)
Comment (Madhu-1) [reply internal/rbd/controllerserver.go:1452]: This is not correct. the vol needs to be destroyed even in case of error.

File: internal/rbd/migration.go (hunk 1 above)
Comment (humblec) [reply internal/rbd/migration.go]: @Madhu-1 you mean along with image name , log rv too?

File: internal/rbd/controllerserver.go (hunk 2 above)
Comment (humblec) [reply internal/rbd/controllerserver.go:1452]: ah... you were referring to defer statement.. that can be brought back.

File: internal/rbd/migration.go (hunk 1 above)
Comment (humblec) [reply internal/rbd/migration.go:86]: This change was not part of , actually I did that as an optimization. I can recheck again..

File: internal/rbd/migration.go (hunk 3 above)
Comment (humblec) [reply internal/rbd/migration.go:83]: This change was not part of , actually I did that as an optimization. I can recheck again..

Comment (humblec) [reply internal/rbd/migration.go:83]: yeah,  reverted it..

File: internal/rbd/migration.go (hunk 1 above)
Comment (humblec) [reply internal/rbd/migration.go]: Done.

**Comment by humblec** [comment]
> if not available Please add e2e changes for this one.

@Madhu-1 you mean in the tracker ?

**Comment by Madhu-1** [comment]
> > if not available Please add e2e changes for this one.
> 
> @Madhu-1 you mean in the tracker ?

as expand is getting adding in this PR it good to add E2E to make sure its working

File: internal/rbd/migration.go [hunk 5]
```diff
@@ -109,3 +96,30 @@ func deleteMigratedVolume(ctx context.Context, parsedMigHandle *migrationVolID,
...
+		return nil, err
+	}
+	// connect to the volume.
+	err = rv.Connect(cr)
+	if err != nil {
+		log.ErrorLog(ctx, "failed to get connected to the rbd image : %s in rv: %v, err: %v", rv.RbdImageName, rv, err)
+
+		return nil, err
+	}
+	defer rv.Destroy()
```
Comment (Madhu-1) [inline internal/rbd/migration.go]: why connect and destory is required here?

File: internal/rbd/controllerserver.go [hunk 6]
```diff
@@ -1447,8 +1447,7 @@ func (cs *ControllerServer) ControllerExpandVolume(
 		return nil, status.Error(codes.InvalidArgument, err.Error())
 	}
 	defer cr.DeleteCredentials()
-
-	rbdVol, err := GenVolFromVolID(ctx, volID, cr, req.GetSecrets())
+	rbdVol, err := genVolFromVolIDWithMigration(ctx, volID, cr, req.GetSecrets())
```
Comment (Madhu-1) [inline internal/rbd/controllerserver.go:1450]: genVolFromVolIDWithMigration return nil in case of err so `defer rbdVol.Destroy()` will panic

**Comment by humblec** [comment]
> > > if not available Please add e2e changes for this one.
> > 
> > 
> > @Madhu-1 you mean in the tracker ?
> 
> as expand is getting adding in this PR it good to add E2E to make sure its working

@Madhu-1 our current e2e for static volumes in general is not upto the mark. so adding an extension again on that at this stage  is not really worth I believe.  Once the sidecars are in place and when we revamp we can add or try to simulate the setups.

File: internal/rbd/rbd_util.go [hunk 7]
```diff
@@ -2071,3 +2071,28 @@ func strategicActionOnLogFile(ctx context.Context, logStrategy, logFile string)
...
+		rv, pErr = genVolFromMigVolID(ctx, pmVolID)
+		if pErr != nil {
+			return nil, pErr
+		}
+	} else {
+		rv, err = GenVolFromVolID(ctx, volID, cr, secrets)
+		rv.Destroy()
+		if err != nil {
+			return nil, err
+		}
```
Comment (Madhu-1) [inline internal/rbd/rbd_util.go]:
```suggestion
		if err != nil {
		rv.Destroy()
			return nil, err
		}
```

and from the caller in case of non error destroy need to be called

File: internal/rbd/rbd_util.go [hunk 8]
```diff
@@ -2071,3 +2071,28 @@ func strategicActionOnLogFile(ctx context.Context, logStrategy, logFile string)
...
+func genVolFromVolIDWithMigration(
+	ctx context.Context, volID string, cr *util.Credentials, secrets map[string]string) (*rbdVolume, error) {
+	var err error
+	var rv *rbdVolume
+	if isMigrationVolID(volID) {
+		pmVolID, pErr := parseMigrationVolID(volID)
+		if pErr != nil {
+			return nil, pErr
+		}
+		rv, pErr = genVolFromMigVolID(ctx, pmVolID)
```
Comment (Madhu-1) [inline internal/rbd/rbd_util.go]: genVolFromMigVolID needs to connect() also or else the resize will fail.

File: internal/rbd/migration.go [hunk 9]
```diff
@@ -77,17 +77,8 @@ func parseMigrationVolID(vh string) (*migrationVolID, error) {
...
-
-	// fill details to rv struct from parsed migration handle
-	rv.RbdImageName = parsedMigHandle.imageName
-	rv.Pool = parsedMigHandle.poolName
-	rv.ClusterID = parsedMigHandle.clusterID
-	rv.Monitors, err = util.Mons(util.CsiConfigFile, rv.ClusterID)
+	rv, err := genVolFromMigVolID(ctx, parsedMigHandle, cr)
 	if err != nil {
-		log.ErrorLog(ctx, "failed to fetch monitors using clusterID: %s, err: %v", rv.ClusterID, err)
-
```
Comment (Madhu-1) [inline internal/rbd/migration.go:90]: Connect is not required in deleteMigratedVolume

File: internal/rbd/migration.go [hunk 10]
```diff
@@ -77,35 +77,43 @@ func parseMigrationVolID(vh string) (*migrationVolID, error) {
 // deleteMigratedVolume get rbd volume details from the migration volID
 // and delete the volume from the cluster, return err if there was an error on the process.
 func deleteMigratedVolume(ctx context.Context, parsedMigHandle *migrationVolID, cr *util.Credentials) error {
+	rv, err := genVolFromMigVolID(ctx, parsedMigHandle, cr)
+	if err != nil {
+		return err
+	}
+	defer rv.Destroy()
+	err = deleteImage(ctx, rv, cr)
```
Comment (Madhu-1) [inline internal/rbd/migration.go:85]:
```suggestion
err = deleteImage(ctx, rv, cr)
if err != nil {
		log.ErrorLog(ctx, "failed to delete rbd image : %s, err: %v", rv, err)
}

return err
}
```

File: internal/rbd/rbd_util.go [hunk 11]
```diff
@@ -2071,3 +2071,28 @@ func strategicActionOnLogFile(ctx context.Context, logStrategy, logFile string)
 		log.ErrorLog(ctx, "unknown cephLogStrategy option %q: hint: 'remove'|'compress'|'preserve'", logStrategy)
 	}
 }
+
+// genVolFromVolIDWithMigration populate a rbdVol structure based on the volID format.
```
Comment (Madhu-1) [inline internal/rbd/rbd_util.go:2075]:
```suggestion
// genVolFromVolIDWithMigration populate a rbdVol structure based on the volID format.
func genVolFromVolIDWithMigration(
	ctx context.Context, volID string, cr *util.Credentials, secrets map[string]string) (*rbdVolume, error) {
	var err error
	var rv *rbdVolume
	if isMigrationVolID(volID) {
		pmVolID, pErr := parseMigrationVolID(volID)
		if pErr != nil {
			return nil, pErr
		}
		rv, pErr = genVolFromMigVolID(ctx, pmVolID, cr)
		if pErr != nil {
		       rv.Destory()
		}
		return rv, pErr
	} 
	rv, err = GenVolFromVolID(ctx, volID, cr, secrets)
	if err != nil {
		rv.Destroy()
	}
	return rv, err
}
```

**Review by Madhu-1 — DISMISSED** [review]
small nit and suggested a nicer version for genVolFromVolIDWithMigration .

**Comment by humblec** [comment]
updated the nicer version too, the  current github action failure is spurious one. `conternarized-test` pass locally.

**Comment by Madhu-1** [comment]
https://github.com/ceph/ceph-csi/pull/2743 should fix CI

File: internal/rbd/migration.go (hunk 1 above)
Comment (Madhu-1) [reply internal/rbd/migration.go]: @humblec its not fixed yet.

File: internal/rbd/rbd_util.go [hunk 12]
```diff
@@ -2071,3 +2071,28 @@ func strategicActionOnLogFile(ctx context.Context, logStrategy, logFile string)
...
+	ctx context.Context, volID string, cr *util.Credentials, secrets map[string]string) (*rbdVolume, error) {
+	var err error
+	var rv *rbdVolume
+	if isMigrationVolID(volID) {
+		pmVolID, pErr := parseMigrationVolID(volID)
+		if pErr != nil {
+			return nil, pErr
+		}
+		rv, pErr = genVolFromMigVolID(ctx, pmVolID, cr)
+		if pErr != nil {
```
Comment (Madhu-1) [inline internal/rbd/rbd_util.go]: sorry for this wrong suggestion Destroy is not needed here. we can just return from here `return genVolFromMigVolID(ctx, pmVolID, cr)`

File: internal/rbd/migration.go [hunk 13]
```diff
@@ -77,35 +77,41 @@ func parseMigrationVolID(vh string) (*migrationVolID, error) {
...
+	if err != nil {
+		log.ErrorLog(ctx, "failed to delete rbd image: %s in rbdVol: %v, err: %v", rv.RbdImageName, rv, err)
```
Comment (Madhu-1) [inline internal/rbd/migration.go]: sorry if i was not clear what i meant was `log.ErrorLog(ctx, "failed to delete rbd image: %s, err: %v", rv, err)` as rv already have String method it will log `pool/radosnamespace/imagename`

**Review by Madhu-1 — DISMISSED** [review]

**Review by Madhu-1 — APPROVED** [review]

**Review by yati1998 — APPROVED** [review]
